### PR TITLE
[MRV2][Spec Decode] Adaptive Speculative Decoding - Initial Support

### PR DESCRIPTION
**Pull request by benchislett**
## Purpose

This PR implements initial support for Adaptive Speculative Decoding using DSpark's confidence head: we enable fully variable-length per request speculative decoding with FULL CUDA Graph support. This PR **does not** implement online dynamic scheduling of the verification budget; instead, we rely on user-provided `num_speculative_tokens_per_batch_size` to determine what size pool of draft tokens to verify at each batch size.

This PR fleshes out a number of fundamental aspects of variable-length speculation, such as:
- Attention metadata flags to declare compatibility with variable length decoding
- FULL CUDA Graph compatibility
- Structured outputs support
- Logprobs support
- Proper integration with statistics capture
- Mixed prefill/decode batch compatibility

## Testing

Since only FLASH_ATTENTION is currently functional, I run with Qwen3-8B-FP8 and the corresponding [DeepSeek-trained DSpark](https://huggingface.co/deepseek-ai/dspark_qwen3_8b_block7). Since verification is so cheap on an 8B model, no significant speedup is measurable; that is not the objective of this PR. However, I do observe an increase in acceptance rate under a fixed verification budget:

<details><summary> Acceptance Rate Study: SPEED-Bench, All Categories, Concurrency 128, Shuffled </summary>

**SPEED-Bench, All Categories, Concurrency 128, Shuffled**

Category | Baseline (K=5) | Baseline (K=7) | Adaptive (Max K=7, Effective K=5 Per Request)
-- | -- | -- | --
coding | 4.6267 | 5.5640 | 5.0972
humanities | 3.2844 | 3.6392 | 3.5036
math | 4.2535 | 5.0100 | 4.5879
multilingual | 3.0081 | 3.4400 | 3.3648
qa | 2.9236 | 3.2078 | 3.0896
rag | 3.7169 | 4.2810 | 4.0475
reasoning | 3.9697 | 4.5843 | 4.2913
roleplay | 2.3373 | 2.4564 | 2.4243
stem | 3.5963 | 4.1574 | 3.9192
summarization | 2.9987 | 3.3062 | 3.1559
writing | 3.1799 | 3.4470 | 3.2988
**Overall Average** | 3.4450 | 3.9176 | 3.7073

Measured using NVIDIA SPEED-Bench harness, thinking off, vLLM config:

```
engine_args:
  speculative_config:
    adaptive_verification: true
    num_speculative_tokens_per_batch_size:
      - [1, 128, 5]
    attention_backend: FLASH_ATTN
  gpu_memory_utilization: 0.75
  attention_backend: FLASH_ATTN
  max_num_batched_tokens: 32768
sampling_kwargs:
  temperature: 0.7
  top_p: 0.8
  top_k: 20
dataset_kwargs:
  shuffle_seed: 42
```

</details>

I also implement various unit tests, including E2E coverage of acceptance length, GSM8k correctness, and an assert that we're observing almost-exactly `N*K` draft tokens at each step.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/attention/backend.py [hunk 1]
```diff
@@ -592,8 +603,11 @@ class AttentionCGSupport(Enum):
     Here we do not consider the cascade attention, as currently
     it is never cudagraph supported."""
 
-    ALWAYS = 3
+    ALWAYS = 4
     """Cudagraph always supported; supports mixed-prefill-decode"""
+    VARLEN_BATCH = 3
+    """Cudagraph supported for variable-length, decode-only batches without
+    consulting host-side sequence or query lengths."""
```
Comment (mgoin) [inline vllm/v1/attention/backend.py]: I don't think this is quite right. For instance this means `AttentionCGSupport.ALWAYS=4` automatically satisfies `VARLEN_BATCH=3`, which is false. IIUC `ALWAYS` describes graphability, while `VARLEN_BATCH` additionally requires the backend not to derive request layout from host-side query lengths. This implies FlexAttention could support this since it returns `ALWAYS`, when it should opt-in.

Given this I think we should make device-driven query metadata an orthogonal capability to AttentionCGSupport, such as `supports_device_query_lengths`.

File: vllm/v1/attention/selector.py [hunk 2]
```diff
@@ -106,6 +117,9 @@ def get_attn_backend(
         use_non_causal=vllm_config.attention_config.use_non_causal,
         use_batch_invariant=envs.VLLM_BATCH_INVARIANT,
         use_kv_connector=use_kv_connector,
+        required_cg_support=(
+            AttentionCGSupport.VARLEN_BATCH if adaptive_verification else None
+        ),
```
Comment (mgoin) [inline vllm/v1/attention/selector.py]: This attention selector runs for both the target and draft model, however I believe this `VARLEN_BATCH` requirement is only required on the target since the drafter is always drafting fixed-shape (per request within a batch size/cudagraph shape). We should only apply this requirement on the target attention selection.

**Comment by jessiewei7** [comment]
re: the scheduler/ModelRunner question — left some notes from SGLang's shipped per-request-K implementation on RFC #48202 (host-side planner emits `verify_lens` as first-class metadata, packed total bucketed to captured graph sizes). fwiw that separation aged well there.

File: vllm/models/deepseek_v4/nvidia/dspark.py [hunk 3]
```diff
@@ -119,6 +120,13 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = "") -> None:
             config.dspark_markov_rank,
             prefix=maybe_prefix(prefix, "markov_head"),
         )
+        self.confidence_head: DSparkConfidenceHead | None = None
+        if vllm_config.speculative_config.adaptive_verification:
+            self.confidence_head = DSparkConfidenceHead(
+                config.hidden_size + config.dspark_markov_rank,
+                prefix=maybe_prefix(prefix, "confidence_head"),
+                params_dtype=vllm_config.model_config.dtype,
+            )
```
Comment (mgoin) [inline vllm/models/deepseek_v4/nvidia/dspark.py:129]: Are you sure bf16 is fine for the confidence head? I see the reference on the model card casts to fp32 to produce fp32 confidence scores https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-DSpark/blob/aa22cb07426656189b2573b8e77a9b7333b8ae0f/inference/model.py#L807-L815

File: vllm/v1/attention/selector.py (hunk 2 above)
Comment (benchislett) [reply vllm/v1/attention/selector.py]: I tried this. It crashed because I think there's no way to have the model runner initialize the attention backend only for its own layers

File: vllm/v1/attention/backend.py (hunk 1 above)
Comment (benchislett) [reply vllm/v1/attention/backend.py]: Yeah this makes sense. I had this before, thought I might be able to simplify and unify into CG requirements. 

Is there any attention backend that actually has one but not the other?

**Comment by benchislett** [comment]
Closing for now, following offline discussion with @LucasWilkinson. Will work on this while he's OOO this/next week.
Iterated design should land somewhere between this PR and https://github.com/vllm-project/vllm/pull/47808

File: vllm/models/deepseek_v4/nvidia/dspark.py (hunk 3 above)
Comment (benchislett) [reply vllm/models/deepseek_v4/nvidia/dspark.py:129]: I would be pretty surprised if FP32 makes a big difference for the confidence head. I guess maybe there's a magnitude issue on the markov rank?

I can double-check the weights, but the outputs can probably be BF16 as we only use them to guide token allocation: being off by even a full percent or two shouldn't make a significant difference

File: tests/v1/e2e/spec_decode/test_spec_decode.py [hunk 4]
```diff
@@ -1434,11 +1434,27 @@ def dspark_config():
             "attention_backend": "FLASH_ATTN",
             "draft_sample_method": "probabilistic",
         },
+        gpu_memory_utilization=0.8,
```
Comment (benchislett) [inline tests/v1/e2e/spec_decode/test_spec_decode.py:1437]: Was getting OOM here due to probabilistic drafting taking up lots of memory for logits, not accounted for by profiler. Not sure if this is safe for CI or not

File: vllm/config/speculative.py [hunk 5]
```diff
@@ -1282,7 +1298,10 @@ def use_dspark(self) -> bool:
         return self.method == "dspark"
 
     def uses_dynamic_speculative_decoding(self) -> bool:
-        return self.num_speculative_tokens_per_batch_size is not None
+        return (
+            self.num_speculative_tokens_per_batch_size is not None
+            and not self.adaptive_verification
```
Comment (benchislett) [inline vllm/config/speculative.py:1303]: Bit of a confusing override here; this is so the scheduler treats adaptive verification as a separate mode (and the model runner decides how many decode tokens to run)

File: vllm/v1/core/sched/scheduler.py [hunk 6]
```diff
@@ -1647,36 +1610,66 @@ def update_from_output(
             scheduled_spec_token_ids = (
                 scheduler_output.scheduled_spec_decode_tokens.get(req_id)
             )
-            # Skip a stale frame still pending discard (async_tokens_to_discard
-            # > 0): its pre-reset rejection count would underflow the counters.
-            if (
-                scheduled_spec_token_ids
-                and (generated_token_ids or self.num_sampled_tokens_per_step == 0)
-                and request.async_tokens_to_discard == 0
+            if scheduled_spec_token_ids and (
```
Comment (benchislett) [inline vllm/v1/core/sched/scheduler.py:1613]: All of this diff is just shuffling things around to fix a bug in the specdec metrics.
Previously we were skipping the metrics observation in cases where `request.async_tokens_to_discard == 0`, and clearing the entry when `num_draft_tokens == 0`.

This overcounts the `num_drafts` since we're only skipping the collection of `num_draft_tokens` and `num_accepted_tokens`, throwing off the scores.

Comment (benchislett) [reply vllm/v1/core/sched/scheduler.py:1613]: Needed for the test where we assert that `num_drafts == K * num_drafted_tokens` in the adaptive SD case.

File: vllm/v1/worker/gpu/model_runner.py [hunk 7]
```diff
@@ -926,10 +950,52 @@ def prepare_inputs(
...
+                    num_reqs + 1, device=self.device, dtype=torch.int32
+                )
+            else:
+                cu_num_logits = async_copy_to_gpu(cu_num_logits_np, device=self.device)
+
         query_start_loc_np = query_start_loc_np[: num_reqs_padded + 1]
         query_start_loc = self.input_buffers.query_start_loc[: num_reqs_padded + 1]
+
+        if draft_tokens:
+            expanded_idx_mapping, expanded_local_pos = expand_idx_mapping(
```
Comment (benchislett) [inline vllm/v1/worker/gpu/model_runner.py:988]: Much of this here is rearranging, since expanded_idx_mapping depends on cu_num_logits, which we construct explicitly in the adaptive verification pathway

File: vllm/v1/worker/gpu/structured_outputs.py [hunk 8]
```diff
@@ -10,7 +10,13 @@
 
 
 class StructuredOutputsWorker:
-    def __init__(self, max_num_logits: int, vocab_size: int, device: torch.device):
+    def __init__(
+        self,
+        max_num_logits: int,
+        vocab_size: int,
+        device: torch.device,
+        mask_stride: int,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/structured_outputs.py:18]: We update structured outputs so it can apply the bitmask to the varlen batch. `mask_stride` effectively makes the bitmask tensor into one that includes padding: and then we read the varlen indices in the kernel, using `position_is_active` to skip any slots for tokens we did not choose to verify

**Comment by cyLi-Tiger** [comment]
One thought on the CUDA graph design: [SGLang's DSpark integration](https://www.lmsys.org/blog/2026-07-06-dspark-sglang) handles per-request variable verify lengths by packing the ragged batch into a compact varlen buffer and keying the graph purely on the total token count. This seems to allow much finer-grained capture tiers and noticeably less padding waste than the approach here, where decode graphs are rounded up to multiples of the scheduled query length. Might be worth considering whether a similar packed-layout design could reduce compute overhead in vLLM's adaptive verify path.

For example, a verify batch [8, 3, 1, 5] (4 reqs, 17 real tokens) with scheduled qlen 8 needs the 32-token FULL graph here (tiers are multiples of 8, and the 24-tier only has 24/8=3 request slots) — 15 padded rows. SGLang's ragged approach packs tokens compactly and keys the graph on total count only, fitting an 18-token tier — 1 padded row, still FULL graph.

cc @benchislett

**Comment by benchislett** [comment]
Closing this PR, we will move forward and try to merge https://github.com/vllm-project/vllm/pull/47808.

**Comment by benchislett** [comment]
Also, @cyLi-Tiger I believe that #47808 does it as you suggest, if I am understanding you properly
